File: 프로그래머스/4/60060. 가사 검색/가사 검색.py
from bisect import bisect_left, bisect_right

array=[[] for _ in range(10001)]
reverse_array=[[] for _ in range(10001)]
def solution(words, queries):
    answer = []


    for word in words:
        array[len(word)].append(word)
        reverse_array[len(word)].append(word[::-1])
    
    for _ in range(10001):
        array[_].sort()
        reverse_array[_].sort()
    
    for query in queries:
        if query[0] == '?' and query[-1] == '?':
            answer.append(len(array[len(query)]))
            continue
        elif query[0]=='?':
            lindex=bisect_left(reverse_array[len(query)], query[::-1].replace('?','a'))
            rindex=bisect_right(reverse_array[len(query)], query[::-1].replace('?','z'))
        else:
            lindex=bisect_left(array[len(query)], query.replace('?','a'))
            rindex=bisect_right(array[len(query)], query.replace('?','z'))
        answer.append(rindex-lindex)
        
    return answer

File: 프로그래머스/4/60060. 가사 검색/가사 검색.cpp
#include <bits/stdc++.h>
using namespace std;

class TrieNode
{
private:
    TrieNode* children[26] = { nullptr };
    int count = 0;

public:
    TrieNode() {}

    void insert(const string& word)
    {
        TrieNode* node = this;
        for (char c : word)
        {
            node->count++;
            int idx = c - 'a';
            if (!node->children[idx])
            {
                node->children[idx] = new TrieNode();
            }
            node = node->children[idx];
        }
    }

    int search(const string& pattern)
    {
        TrieNode* node = this;
        for (char c : pattern)
        {
            if (c == '?') return node->count;

            int idx = c - 'a';
            if (!node->children[idx]) return 0;
            node = node->children[idx];
        }
        return 0;
    }
};

TrieNode forwardTrie[10001];
TrieNode reverseTrie[10001];

vector<int> solution(vector<string> words, vector<string> queries)
{
    vector<int> answer;

    for (const string& word : words)
    {
        int len = word.length();
        forwardTrie[len].insert(word);

        string rev = word;
        reverse(rev.begin(), rev.end());
        reverseTrie[len].insert(rev);
    }

    for (const string& q : queries)
    {
        int len = q.length();
        if (q[0] != '?')
        {
            answer.push_back(forwardTrie[len].search(q));
        }
        else
        {
            string revQuery = q;
            reverse(revQuery.begin(), revQuery.end());
            answer.push_back(reverseTrie[len].search(revQuery));
        }
    }

    return answer;
}
